kythe/go/serving/tools/kythefs/kythefs.go

```diff
@@ -0,0 +1,401 @@
+			}
+			if isDir {
+				return &fuse.Attr{
+					Mode: fuse.S_IFDIR | 0755,
+				}, fuse.OK
+			} else {
+				src, err := me.FetchSourceForURI(*r.KytheURI)
+				if err != nil {
+					return nil, fuse.ENOENT
+				}
+				return &fuse.Attr{
+					Mode: fuse.S_IFREG | 0644, Size: uint64(len(src)),
+				}, fuse.OK
+			}
+		}
+	}
+	return nil, fuse.ENOENT
+}
+
+func (me *KytheFs) OpenDir(path string, context *fuse.Context) (c []fuse.DirEntry, code fuse.Status) {
+	resolutions, err := me.ResolveFilepath(path)
+
+	if err != nil {
+		log.Printf("Resolution error for [%v]: %v", path, err)
+		return nil, fuse.ENOENT
+	}
+
+	// Key by path component, since when a corpus+root segment overlaps with
+	// an actual deep path from an other corpus+root, we could get duplicate
+	// components otherwise.
+	ents := make(map[string]fuse.DirEntry)
+	for _, r := range resolutions {
+		if r.NextDirComponent != "" {
+			ents[r.NextDirComponent] = fuse.DirEntry{
+				Name: r.NextDirComponent,
+				Mode: fuse.S_IFDIR,
+			}
+		} else if r.KytheURI != nil {
+			req := &ftpb.DirectoryRequest{
+				Corpus: r.KytheURI.Corpus,
+				Root:   r.KytheURI.Root,
+				Path:   r.KytheURI.Path,
+			}
+			dir, err := me.FileTree.Directory(me.Context, req)
+
+			if err != nil {
+				log.Printf("Error fetching dir contents for [%v] (ticket [%v]): %v",
+					path, r.KytheURI.String(), err)
+				return nil, fuse.ENOENT
+			}
+
+			for _, d := range dir.Subdirectory {
+				uri, err := kytheuri.Parse(d)
+				if err != nil {
+					log.Printf("warning: received invalid directory uri %v: %v", d, err)
+					continue
+				}
+				component := filepath.Base(uri.Path)
+				ents[component] = fuse.DirEntry{
+					Name: component,
+					Mode: fuse.S_IFDIR,
+				}
+			}
+			for _, f := range dir.File {
+				uri, err := kytheuri.Parse(f)
+				if err != nil {
+					log.Printf("warning: received invalid file uri %v: %v", f, err)
+					continue
+				}
+				component := filepath.Base(uri.Path)
+				ents[component] = fuse.DirEntry{
+					Name: component,
+					Mode: fuse.S_IFREG,
+				}
+			}
+		} else {
+			log.Fatalf(
+				"Programming error: resoultion is neither dir part nor uri for [%v]",
+				path)
+		}
+	}
+	var result []fuse.DirEntry
+	for _, v := range ents {
+		result = append(result, v)
+	}
+	return result, fuse.OK
+}
+
+func (me *KytheFs) Open(path string, flags uint32, context *fuse.Context) (file nodefs.File, code fuse.Status) {
+	// Read-only filesystem.
+	if flags&fuse.O_ANYWRITE != 0 {
+		return nil, fuse.EPERM
+	}
+
+	src, err := me.FetchSource(path)
+	if err != nil {
+		log.Printf("Error fetching source for [%v]: %v", path, err)
+		return nil, fuse.ENOENT
+	}
+
+	return nodefs.NewDataFile(src), fuse.OK
+}
+
+//
+// Main
+//
+
+func main() {
+	flag.Parse()
+	if *serverAddr == "" {
+		log.Fatal("You must provide --server address")
+	}
+	if *mountPoint == "" {
+		log.Fatal("You must provide --mountpoint")
+	}
+
+	var xrefClient xrefs.Service = xrefs.WebClient(*serverAddr)
```

Comment:
```
xs := xrefs.WebClient(*serverAddr)
ft := filetree.WebClient(*serverAddr)
gs := graph.WebClient(*serverAddr)
```

Or even better, use https://godoc.org/kythe.io/kythe/go/serving/api#ParseSpec

Reply:
Ah, it took me a while to figure what nameless fields meant in a type.. Thank you.